include/domain-types.h

```diff
@@ -121,6 +121,28 @@ struct Tag
 
 typedef std::vector<Tag> Tags;
 
+struct Payload
+{
+    Payload(const std::string& fileName)
+        : fileName(fileName)
+        , data(nullptr)
+        , dataSize(0)
+    {
+    }
+
+    Payload(const void* data, size_t dataSize, const std::string& mimeType)
+        : data(data)
+        , dataSize(dataSize)
+        , mimeType(mimeType)
+    {
+    }
+
+    std::string fileName;
+    const void* data;
```

Comment:
To be clean implementation, then your caller shall create
shared_array pointer (instead of std::vector) and your object shall contain shared_array pointer. Then you have no copying and no pointers with unknown ownership.